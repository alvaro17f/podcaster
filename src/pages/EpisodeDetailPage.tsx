import { useEffect, useState } from "react";
import { getLocalStorage, setLocalStorage } from "../helpers/localStorage";
import { PodcastDetail } from "../components/PodcastDetail";
import { useParams } from "react-router-dom";
import { EpisodeDetail } from "../components/EpisodeDetail";

export const EpisodeDetailPage = () => {
	const [podcasts, setPodcasts] = useState();
	const [episodes, setEpisodes] = useState();
	const [error, setError] = useState();
	const [loadingPodcasts, setLoadingPodcasts] = useState(true);
	const [loadingEpisodes, setLoadingEpisodes] = useState(true);

	const params = useParams();

	useEffect(() => {
		setLoadingPodcasts(true);

		if (getLocalStorage("podcasts")) {
			setPodcasts(getLocalStorage("podcasts"));
			setLoadingPodcasts(false);
		} else {
			fetch(
				`https://api.allorigins.win/raw?url=${encodeURIComponent(
					"https://itunes.apple.com/us/rss/toppodcasts/limit=100/genre=1310/json",
				)}`,
			)
				.then((response) => {
					if (response.ok) return response.json();
					throw new Error("Network response was not ok.");
				})
				.then((data) => {
					setPodcasts(data);
					setLocalStorage("podcasts", data, 86400000);
				})
				.catch((e) => {
					console.log(e);
					setError(e);
				})
				.finally(() => setLoadingPodcasts(false));
		}
	}, []);

	useEffect(() => {
		setLoadingEpisodes(true);

		if (getLocalStorage(`podcastDetail${params.podcastId}`)) {
			setEpisodes(getLocalStorage(`podcastDetail${params.podcastId}`));
			setLoadingEpisodes(false);
		} else {
			fetch(
				`https://api.allorigins.win/raw?url=${encodeURIComponent(
					`https://itunes.apple.com/lookup?id=${params.podcastId}&media=podcast&entity=podcastEpisode`,
				)}`,
			)
				.then((response) => {
					if (response.ok) return response.json();
					throw new Error("Network response was not ok.");
				})
				.then((data) => {
					setEpisodes(data);
					setLocalStorage(`podcastDetail${params.podcastId}`, data, 86400000);
				})
				.catch((e) => {
					console.log(e);
					setError(e);
				})
				.finally(() => setLoadingEpisodes(false));
		}
	}, []);

	if (error) {
		return (
			<>
				<div className="grid text-5xl place-items-center">ERROR</div>;
				<pre className="grid mt-5 place-items-center">{error}</pre>
			</>
		);
	}
	return (
		<>
			<section className="grid grid-cols-[1fr_3fr] gap-20 m-5">
				<div className="shadow-xl span-2">
					{loadingPodcasts ? (
						<p>LOADING...</p>
					) : (
						<PodcastDetail data={podcasts?.feed.entry} id={params.podcastId} />
					)}
				</div>
				<div className="grid pl-5 text-2xl shadow-xl">
					{loadingEpisodes ? (
						<p>LOADING...</p>
					) : (
						<EpisodeDetail data={episodes?.results} id={params.episodeId} />
					)}
				</div>
			</section>
		</>
	);
};
