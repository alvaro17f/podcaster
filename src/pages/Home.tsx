import { useEffect, useState } from "react";
import { getLocalStorage, setLocalStorage } from "../helpers/localStorage";
import { Podcast } from "../components/Podcast";

export const Home = () => {
	const [data, setData] = useState();
	const [error, setError] = useState();
	const [loading, setLoading] = useState(true);
	const [query, setQuery] = useState("");

	const search = (data) => {
		return data.filter(
			(data) =>
				data["im:artist"].label.toLowerCase().includes(query) ||
				data.title.label.toLowerCase().includes(query),
		);
	};

	useEffect(() => {
		setLoading(true);

		if (getLocalStorage("data")) {
			setData(getLocalStorage("data"));
			setLoading(false);
		} else {
			fetch(
				`https://api.allorigins.win/raw?url=${encodeURIComponent(
					"https://itunes.apple.com/us/rss/toppodcasts/limit=100/genre=1310/json",
				)}`,
			)
				.then((response) => {
					if (response.ok) return response.json();
					throw new Error("Network response was not ok.");
				})
				.then((data) => {
					setData(data);
					setLocalStorage("data", data, 86400000);
				})
				.catch((e) => {
					console.log(e);
					setError(e);
				})
				.finally(() => setLoading(false));
		}
	}, []);

	if (loading) {
		return <div className="grid text-5xl place-items-center">LOADING...</div>;
	}
	if (error) {
		return (
			<>
				<div className="grid text-5xl place-items-center">ERROR</div>;
				<pre className="grid mt-5 place-items-center">{error}</pre>
			</>
		);
	}

	return (
		<>
			<div className="flex justify-end gap-3 pt-5 pr-5 flex-nowrap">
				<div className="p-2 text-3xl text-white bg-blue-400 rounded-xl">
					{data?.feed.entry.length}
				</div>
				<input
					type="search"
					name="search"
					placeholder="Filter podcasts..."
					className="p-2 border border-gray-200 rounded-xl focus:outline-gray-300"
					onChange={({ target }) => setQuery(target.value.toLocaleLowerCase())}
				/>
			</div>

			<div className="grid grid-cols-4 gap-10 mx-10 mt-60 place-items-center">
				{search(data?.feed.entry).map((data, idx) => (
					<Podcast key={idx} data={data} />
				))}
			</div>
		</>
	);
};
