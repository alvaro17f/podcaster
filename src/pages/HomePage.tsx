import { useEffect, useState } from "react";
import { getLocalStorage, setLocalStorage } from "../utils/localStorage";
import { Podcast } from "../components/Podcast";
import { awesomeReducer } from "../hooks/awesomeReducer";
import { useFetch } from "../hooks/useFetch";


type Props = {
  "im:image": {
    label: string;
  }[];
  "im:artist": {
    label: string;
  };
  id: {
    attributes: {
      "im:id": string;
    };
  };
  title: {
    label: string;
  };
}[];

export const HomePage = () => {
  const { ENUMS, dispatch, state } = awesomeReducer()
  const [query, setQuery] = useState("");

  const search = (podcasts: Props) => {
    return podcasts?.filter(
      (podcasts) =>
        podcasts["im:artist"].label.toLowerCase().includes(query) ||
        podcasts.title.label.toLowerCase().includes(query),
    );
  };

  const { data: podcastData, setShouldFetch } = useFetch(
    `https://api.allorigins.win/raw?url=${encodeURIComponent(
      "https://itunes.apple.com/us/rss/toppodcasts/limit=100/genre=1310/json",
    )}`,
  );

  useEffect(() => {
    if (getLocalStorage("podcasts")) {
      dispatch({ type: ENUMS.SET_PODCASTS, payload: getLocalStorage("podcasts") });
    } else {
      setShouldFetch(true);
      podcastData && (
        dispatch({ type: ENUMS.SET_PODCASTS, payload: podcastData }),
        setLocalStorage("podcasts", podcastData, 86400000))
    }
  }, [podcastData]);


  if (state.loadingPodcasts) {
    return <div className="grid text-5xl place-items-center">LOADING...</div>;
  }
  if (state.error) {
    return <div className="grid text-5xl place-items-center">ERROR</div>;
  }

  return (
    <section>
      <div className="flex justify-end gap-3 pt-5 pr-5 flex-nowrap">
        <div className="p-2 text-3xl text-white bg-blue-400 rounded-xl">
          {state.podcasts?.feed.entry.length}
        </div>
        <input
          type="search"
          name="search"
          placeholder="Filter podcasts..."
          className="p-2 border border-gray-200 rounded-xl focus:outline-gray-300"
          onChange={({ target }) => setQuery(target.value.toLocaleLowerCase())}
        />
      </div>

      <div className="grid grid-cols-4 gap-10 mx-10 mt-60 place-items-center">
        {search(state.podcasts?.feed.entry)?.map((podcasts, idx) => (
          <Podcast key={idx} podcasts={podcasts} />
        ))}
      </div>
    </section>
  );
};
