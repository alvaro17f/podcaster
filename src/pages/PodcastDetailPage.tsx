import { useEffect, useState } from "react";
import { getLocalStorage, setLocalStorage } from "../utils/localStorage";
import { PodcastDetail } from "../components/PodcastDetail";
import { useParams } from "react-router-dom";
import { Table } from "../components/Table";
import { useFetch } from "../hooks/useFetch";
import { awesomeReducer } from "../hooks/awesomeReducer";

export const PodcastDetailPage = () => {
  const { ENUMS, dispatch, state } = awesomeReducer();
  const params = useParams();

  const { data: podcastData, setShouldFetch: fetchPodcasts } = useFetch(
    `https://api.allorigins.win/raw?url=${encodeURIComponent(
      "https://itunes.apple.com/us/rss/toppodcasts/limit=100/genre=1310/json",
    )}`,
  );

  const { data: episodeData, setShouldFetch: fetchEpisodes } = useFetch(
    `https://api.allorigins.win/raw?url=${encodeURIComponent(
      `https://itunes.apple.com/lookup?id=${params.podcastId}&media=podcast&entity=podcastEpisode`,
    )}`,
  );

  useEffect(() => {
    try {
      if (getLocalStorage("podcasts")) {
        dispatch({
          type: ENUMS.SET_PODCASTS,
          payload: getLocalStorage("podcasts"),
        });
      } else {
        fetchPodcasts(true);
        podcastData && (
          dispatch({ type: ENUMS.SET_PODCASTS, payload: podcastData }),
          setLocalStorage("podcasts", podcastData, 86400000)
        )
      }

      if (getLocalStorage(`podcastDetail${params.podcastId}`)) {
        dispatch({
          type: ENUMS.SET_EPISODES,
          payload: getLocalStorage(`podcastDetail${params.podcastId}`),
        });
      } else {
        fetchEpisodes(true);
        episodeData && (
          dispatch({ type: ENUMS.SET_EPISODES, payload: episodeData }),
          setLocalStorage(
            `podcastDetail${params.podcastId}`,
            episodeData,
            86400000,
          )
        );
      }
    } catch (e) {
      dispatch({
        type: ENUMS.SET_ERROR,
        payload: e,
      });
      console.log(e);
    }
  }, [podcastData, episodeData]);

  if (state.error) {
    return <div className="grid text-5xl place-items-center">ERROR</div>;
  }
  return (
    <section className="grid grid-cols-[1fr_3fr] gap-20 m-5">
      <div className="shadow-xl h-fit">
        {state.loadingPodcasts ? (
          <p>LOADING...</p>
        ) : (
          <PodcastDetail
            data={state.podcasts?.feed.entry}
            id={params.podcastId}
          />
        )}
      </div>
      <div>
        <div className="flex items-center h-16 pl-5 text-2xl font-bold shadow-xl">
          {state.loadingEpisodes ? (
            <p>LOADING...</p>
          ) : (
            <p>EPISODES: {state.episodes?.results.length}</p>
          )}
        </div>

        <div className="text-2xl shadow-xl p-7">
          {state.loadingEpisodes ? (
            <p>LOADING...</p>
          ) : (
            <Table data={state.episodes?.results} id={params.podcastId} />
          )}
        </div>
      </div>
    </section>
  );
};
