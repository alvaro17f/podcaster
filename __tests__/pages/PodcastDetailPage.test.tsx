import React from "react";
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import {
  render,
  screen,
  cleanup,
  waitFor,
} from "@testing-library/react";
import { PodcastDetailPage } from "../../src/pages/PodcastDetailPage";
import { MemoryRouter, Route, Routes } from "react-router-dom";

const wrapper = (
  <MemoryRouter initialEntries={["/podcast/1535809341"]}>
    <Routes>
      <Route path="/podcast/:podcastId" element={<PodcastDetailPage />} />
    </Routes>
  </MemoryRouter>
);

describe("<PodcastDetailPage />", () => {
  beforeEach(() => {
    render(wrapper);
  });
  afterEach(cleanup);

  it("should match the snapshot", () => {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });

  it("should render 'LOADING...'", () => {
    screen.getAllByText(/loading/i);
  });

  it("should render EPISODES", async () => {
    await waitFor(() => {
      screen.getAllByText(/episodes/i);
    });
  });

  it("should render 1 image", async () => {
    await waitFor(() => {
      expect(screen.getAllByRole("img")).toHaveLength(1);
      screen.getByAltText(
        "The Joe Budden Podcast - The Joe Budden Network picture",
      );
    });
  });

  it("should render description", async () => {
    await waitFor(() => {
      screen.getByText(/description/i);
      screen.getByText(
        "Tune into Joe Budden and his friends. Follow along the crazy adventures of these very random friends.",
      );
    });
  });

  it("should render a episode", async () => {
    await waitFor(() => {
      screen.getByRole("link", { name: /Episode 618 | "They Got Us"/i });
      screen.getByText("15/4/2023");
      screen.getByText("177:53");
    });
  });

});
