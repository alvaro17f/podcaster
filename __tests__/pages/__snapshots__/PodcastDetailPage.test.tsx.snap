// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PodcastDetailPage /> > should match the snapshot 1`] = `
<div>
  <section
    class="grid grid-cols-[1fr_3fr] gap-20 m-5"
  >
    <div
      class="shadow-xl h-fit"
    >
      <p>
        LOADING...
      </p>
    </div>
    <div>
      <div
        class="flex items-center h-16 pl-5 text-2xl font-bold shadow-xl"
      >
        <p>
          LOADING...
        </p>
      </div>
      <div
        class="text-2xl shadow-xl p-7"
      >
        <p>
          LOADING...
        </p>
      </div>
    </div>
  </section>
</div>
`;
